### ISSUE TITLE
Implement advanced collections

### ISSUE THREAD

**Issue by natebrennand** [natebrennand/cpp_collections#3]
- [x] fold
- [x] reduce
- [x] foldLeft
- [x] foldRight
- [x] reduceRight
- [x] reduceLeft
- [x] scanLeft
- [x] scanRight

**Comment by ianzapolsky** [comment]
For take, it seems like we are going to have to create a way for users to define collections based on a step function. like a generator

**Comment by natebrennand** [comment]
Hmm, we'll have to think about how we want to do that exactly.
I guess maybe a function that takes in `(int index, T last)` and generates the next?

**Comment by ianzapolsky** [comment]
It seems like that wouldn't work for something like the Fibonacci sequence

**Comment by natebrennand** [comment]
hmmm, you're right.
So if we're possible to infer how many arguments a function takes, we might be able to pass the last N values back to the function?

That seems really hacky though and I think it'd be a difficult interface to explain even if we got it functioning.

**Comment by natebrennand** [comment]
hmm, so fibonacci is still possible if you define a functor, right? You would just need a stateful object, which is not a great design but it is a possible way to do this I think

**Comment by ianzapolsky** [comment]
I agree. This concept is the most challenging for me to wrap my head around. 

If we are going to do 'take', we want the user to be able to define non-trivial generators, which could fall under the category of just a simple increment, or a much more complex function or combination of other collections.

**Comment by natebrennand** [comment]
Another way is to pass in a function that returns `fn(U val) -> pair<T, U>`. So you give it a seed (in the case of fibonacci, tuple(1, 1)), then you return a value and the next seed every time you call the fn.

**Comment by ianzapolsky** [comment]
Where does the state live though? Inside the user defined functor object that he passes to the generator?

**Comment by natebrennand** [comment]
```cpp
auto fib = [](tuple<int, int> prev) {
    int next = get<0>(prev) + get<1>(prev);
    return make_pair(next, make_tuple(get<1>(prev), next));
}
```

-------
re: functor, state would live within the object. I don't think that's a good way to go about it though.

**Comment by ianzapolsky** [comment]
What would the alternative be to storing internal state? The upside of the example you provide is they pretty much any pattern can be defined by the user if we make these functions variadic

**Comment by natebrennand** [comment]
Oh sorry, I meant that I don't think a functor is a good way to go about this.
I think the other strategy is a lot cleaner and would be easier to use than functors.

**Comment by ianzapolsky** [comment]
Also would we be able to achieve this kind of recursive sequence definition? (Haskell) fibs = 0 : 1 : zipWith (+) fibs (tail fibs)

**Comment by ianzapolsky** [comment]
Oh ok , yeah. To answer my own question I suppose this would work as long as we pass the collection fibs through the tuple to the functor

**Comment by ianzapolsky** [comment]
(And we change the zipWith implementation to set result size equal to that of the larger list passed to it)

**Comment by natebrennand** [comment]
I'm not entirely following your last 3 comments. I don't know enough haskell syntax and I'm not sure what you're responding to. (the `>` command works for quoting in markdown)

Why would `zipWith` default to the larger size? What kind of value would it use for the tail end values of the shorter list?

**Comment by ianzapolsky** [comment]
sorry, wrote those comments from my phone. basically, the way the above haskell snippet defines the list called fibs is recursively, using zipWith. tail returns everything except the first element of the list. so the best way to see what's going on is with a [picture](http://stackoverflow.com/questions/6273621/understanding-a-recursively-defined-list-fibs-in-terms-of-zipwith)

**Comment by ianzapolsky** [comment]
by the way, at the moment, we have implemented foldLeft and reduceLeft, and named them fold and reduce. For the sake of completeness it seems like we should have fold, reduce, and scan implemented for both left and right direction

**Comment by natebrennand** [comment]
That sounds fine. Though I don't fully understand the use cases for {fold,reduce}Right

**Comment by natebrennand** [comment]
closed by #30
